ddl/index.go

```diff
@@ -883,12 +886,79 @@ func (w *worker) addTableIndex(t table.Table, indexInfo *model.IndexInfo, reorgI
 	}
 	defer closeAddIndexWorkers(idxWorkers)
 
-	kvRanges, err := splitTableRanges(t, reorgInfo)
+	finish := false
+	for !finish {
+		kvRanges, err := splitTableRanges(t, reorgInfo)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		err = w.buildKVRangesIndex(t, idxWorkers, kvRanges, job, reorgInfo)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		finish, err = w.updateReorgInfo(t, reorgInfo)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+// updateReorgInfo will find the next partition according to current reorgInfo.
+func (w *worker) updateReorgInfo(t table.Table, reorg *reorgInfo) (bool, error) {
+	pi := t.Meta().GetPartitionInfo()
+	if pi == nil {
+		return true, nil
+	}
+
+	pid, err := findNextPartitionID(reorg.PartitionID(), pi.Definitions)
 	if err != nil {
-		return errors.Trace(err)
+		// Fatal error, should not run here.
+		return false, errors.Errorf("wrong reorgInfo, partition id %d not found", reorg.PartitionID())
+	}
+	if pid == -1 {
+		return true, nil
+	}
+
+	start, end, err := getPartitionRange(reorg.d, t.Meta(), pid, reorg.Job.SnapshotVer)
+	if err != nil {
+		return false, errors.Trace(err)
 	}
 
-	return w.buildKVRangesIndex(t, idxWorkers, kvRanges, job, reorgInfo)
+	reorgMeta := model.NewDDLReorgMeta()
+	reorgMeta.PartitionID = pid
+	reorgMeta.EndHandle = end
+	reorg.Job.ReorgMeta = reorgMeta
+	reorg.StartHandle = start
+
+	log.Infof("[ddl] job %v update reorgInfo partitionID:%v, startHandle:%v, endHandle:%v", reorg.Job.ID, pid, start, end)
+
+	// Actually, it just need to update the job.ReorgMeta.
+	err = kv.RunInNewTxn(reorg.d.store, false, func(txn kv.Transaction) error {
+		t := meta.NewMeta(txn)
+		return errors.Trace(w.updateDDLJob(t, reorg.Job, false))
+	})
+	return false, errors.Trace(err)
+}
+
+// findNextPartitionID finds the next partition ID in the PartitionDefinition.
+// Returns -1 if this partitionID is the last one.
+func findNextPartitionID(partitionID int64, defs []model.PartitionDefinition) (int64, error) {
+	idx := -1
+	for i, def := range defs {
+		if partitionID == def.ID {
+			idx = i
```

Comment:
Why not use "ID" directly?

Reply:
What do you mean by "ID" ?

Reply:
See the comment of this function:
> // findNextPartitionID finds the next partition ID in the PartitionDefinition.
> // Returns -1 if this partitionID is the last one.

This function find the **next** partition ID, using offset is more convenient. @zimulala

Reply:
Could we write it as follows code?
```
if partitionID == def.ID {
    if i == len(defs)-1{
        return 0
    }else{
        return def[i+1].ID
    }
}
```